drv/lpc55-update-server/src/images.rs

```diff
@@ -23,82 +31,569 @@ extern "C" {
-                __IMAGE_A_BASE.as_ptr() as u32..__IMAGE_A_END.as_ptr() as u32,
-            ),
-            (RotComponent::Hubris, SlotId::B) => (
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+            (RotComponent::Hubris, SlotId::A) => FlashRange {
+                store: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+                exec: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+            },
+            (RotComponent::Hubris, SlotId::B) => FlashRange {
+                store: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+                exec: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+            },
+            (RotComponent::Stage0, SlotId::A) => FlashRange {
+                store: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
+            },
+            (RotComponent::Stage0, SlotId::B) => FlashRange {
+                store: __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+            },
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC format.
+impl From<&[u8]> for Epoch {
+    fn from(chars: &[u8]) -> Self {
+        // Ten digits are necessary to represent u32::MAX
+        if chars.len() != 10 {
+            return Epoch { value: 0 };
+        }
+        // Ten ASCII decimal digits can represent a number
+        // greater than u32::MAX.
+        match chars.iter().try_fold(0u32, |epoch: u32, c| {
+            if c.is_ascii_digit() {
+                epoch.checked_mul(10).and_then(|product| {
+                    product.checked_add(c.saturating_sub(b'0') as u32)
+                })
+            } else {
+                None
+            }
+        }) {
+            Some(value) => Epoch { value },
+            None => Epoch::from(0),
+        }
+    }
+}
```

Comment:
This seems like it's reimplementing https://doc.rust-lang.org/stable/core/primitive.u32.html#method.from_str_radix

Reply:
I had the EPOC format as 10 ASCII decimal digits and that I would be checking format and comparing u8's.
But, since adding support for using the u32 from ImageHeader for early checking during update, the EPOC value from the caboose needs to be converted to a u32 and the leading zeros don't add any value any more. I'm removing the requirement for the EPOC value being a [u8; 10].
`from_str_radix()` has a call to panic if the radix is outside 2..=36. That probably gets optimized out since the radix is a constant. It also requires a &str which means the &[u8] needs to be legal utf8. This is foreign data (pre-signature check), so we cannot avoid the check without risking UB.
This explicit conversion makes it easier, for me at least, to reason about the code.

Reply:
Done in impl From<&[u8]> for Epoch { fn from(chars: &[u8]) -> Self { ... chars.parse::<u32>(). ... }